File: PlaylistObjC/PlaylistObjC/PlaylistDetailViewController.m
//
//  PlaylistDetailViewController.m
//  PlaylistObjC
//
//  Created by James Pacheco on 4/21/16.
//  Copyright © 2016 DevMountain. All rights reserved.
//

#import "PlaylistDetailViewController.h"
#import "PlaylistController.h"

@interface PlaylistDetailViewController ()

@property (nonatomic, weak) IBOutlet UITextField *titleTextField;
@property (nonatomic, weak) IBOutlet UITextField *artistTextField;
@property (nonatomic, weak) IBOutlet UITableView *tableView;

@end

@implementation PlaylistDetailViewController

#pragma mark - Actions

- (IBAction)addButtonTapped:(id)sender
{
	[[PlaylistController sharedInstance] addSongWithTitle:self.titleTextField.text andArtist:self.artistTextField.text toPlaylist:self.playlist];
	[self.tableView reloadData];
}

#pragma mark - UITableViewDataSource/Delegate

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return [self.playlist.songs count];
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"songCell" forIndexPath:indexPath];
    Song *song = self.playlist.songs[indexPath.row];
    cell.textLabel.text = song.title;
    cell.detailTextLabel.text = song.artist;
    return cell;
}

- (void)tableView:(UITableView *)tableView commitEditingStyle:(UITableViewCellEditingStyle)editingStyle forRowAtIndexPath:(NSIndexPath *)indexPath
{
    if (editingStyle == UITableViewCellEditingStyleDelete) {
        Song *song = self.playlist.songs[indexPath.row];
        [[PlaylistController sharedInstance] deleteSong:song fromPlaylist:self.playlist];
        [self.tableView deleteRowsAtIndexPaths:@[indexPath] withRowAnimation:UITableViewRowAnimationAutomatic];
    }
}

#pragma mark - Properties

- (void)setPlaylist:(Playlist *)playlist
{
	if (playlist != _playlist) {
		_playlist = playlist;
		[self.tableView reloadData];
	}
}

@end

File: Playlist/CoreDataMiniProject/PlaylistDetailViewController.swift
//
//  PlaylistDetailViewController.swift
//  CoreDataMiniProject
//
//  Created by Jordan Nelson on 2/15/16.
//  Copyright © 2016 Jordan Nelson. All rights reserved.
//

import UIKit

class PlaylistDetailViewController: UIViewController {
    
    @IBOutlet weak var tableView: UITableView!
    
    @IBOutlet weak var titleField: UITextField!
    @IBOutlet weak var artistField: UITextField!
    
    var playlist: Playlist?
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        if let playlist = playlist {
            updateViewWithPlaylist(playlist)
        }
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    func updateViewWithPlaylist(playlist:Playlist){
        title = playlist.name
    }
    
    @IBAction func addButtonTapped(sender: AnyObject) {
        if let playlist = playlist,
            title = titleField.text,
            artist = artistField.text {
            PlaylistController.sharedController.addSongToPlaylist(title, artist: artist, playlist: playlist)
            tableView.reloadData()
        }
    }
}

extension PlaylistDetailViewController: UITableViewDelegate, UITableViewDataSource {
    
    func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCellWithIdentifier("songCell", forIndexPath: indexPath)
        if let songs = playlist?.songs {
            let song = songs[indexPath.row]
            
            cell.textLabel?.text = song.title
            cell.detailTextLabel?.text = song.artist
        }
        
        return cell
    }
    
    func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        if let songs = playlist?.songs {
            return songs.count
        } else {
            return 0
        }
    }
    
    func tableView(tableView: UITableView, commitEditingStyle editingStyle: UITableViewCellEditingStyle, forRowAtIndexPath indexPath: NSIndexPath) {
        guard let playlist = playlist else {return}
        if editingStyle == .Delete {
            let song = playlist.songs[indexPath.row]
            
            PlaylistController.sharedController.removeSongFromPlaylist(song, playlist: playlist)
            tableView.deleteRowsAtIndexPaths([indexPath], withRowAnimation: .Automatic)
        }

    }
}



